### PR TITLE
Implement serialization for transform functions.

### LINKED ISSUE #15194: Implement serialization of transform functions
Currently, only serialization of `translate()` function is implmeneted. We should have the rest implemented as well.

Spec: https://drafts.csswg.org/css-transforms/#serialization-of-transform-functions

The relevant code is https://github.com/servo/servo/blob/master/components/style/properties/longhand/box.mako.rs#L1183-L1237

(I think this is an easy one, but not sure. cc @Manishearth)

### PR DESCRIPTION
**Pull request by pyfisch**
Preserve more information from transform function parsing.
Preserve angle unit while parsing.
Simplify SpecifiedMatrix.
Use the write! macro for formatting with a helper called Css.
Implement ToCss for &T if T implements ToCss.
Add some tests and update others.

closes #15194

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15194 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16242)

### CODE REVIEW

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: Please ensure to keep the value finite, since there's other code that relies on this. Please adjust the constructors appropriately, and also the `radians` method.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: I guess this interacts with #16231... Oh well :(

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Let's keep the `mij` convention for the matrix members?

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Let's move this to `style_traits`, since it may be helpful for other consumers. Also, let's make that comment a doc comment.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: nit: Indentation is off here.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: This needs to be `input.try(|input| input.expect_comma())` instead, otherwise you're accepting stuff like `translate(2px foo)`, and parsing it as `translate(2px)`.

Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Please add a test for this.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: ditto.

Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Also, let's use just `if input.try(|input| input.expect_comma()).is_ok()` instead of `if let`, for consistency with all the other existing code.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: ditto.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: nit: Let's be consistent and remove the `SpecifiedOperation::` prefix from the rest of the branches.

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: nit: Let's just implement `AngleUnit::to_css` here, and call it from here instead.

**Review by emilio — CHANGES_REQUESTED** [review]
Looks generally ok! Did a first pass, thanks for working on this :)

File: components/style/values/specified/mod.rs
Comment (pyfisch) [inline components/style/values/specified/mod.rs]: I add a check to the `radians()` method.

File: components/style/properties/longhand/box.mako.rs
Comment (pyfisch) [inline components/style/properties/longhand/box.mako.rs]: The spec already calls the matrix members a, b, c, d, e, f:
https://drafts.csswg.org/css-transforms/#MatrixDefined

**Comment by pyfisch** [comment]
@emilio applied requested changes and fixed nits.

**Comment by Manishearth** [comment]
r? @emilio since he already started review.

@canaltinova , might want to have a look at this too

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: Looks reasonable then, thanks

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: nit: a single `Transform Function` (without the plural).

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: nit: brace and paren on the next line.

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: nit: you could also use `self.value.to_css(dest)` if you preferred, it'd desugar to the same though, so your call.

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: I _think_ we may still need to prevent infinite values from coming here, because Gecko may rely on it even for specified values. I'll do a Gecko try run to double-check though.

**Review by emilio — APPROVED** [review]
Looks good to me, only a few nits left. I'll do a try run on Servo's and Gecko's infra while the last nits are addressed, and land this if nobody else raises any concern.

Thanks a bunch for working on this! \o/

**Comment by emilio** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=ecb364953429bc19871e35871dd2a7d9ae2d084f

File: components/style/values/specified/mod.rs
Comment (pyfisch) [inline components/style/values/specified/mod.rs]: The only way to access the value of angle is through `radians` and it is checked there.

It may be a problem if NaN is passed to Angle::from_{radians, degrees, ...} and the angle is serialized with ToCss. Is this problematic or not?

File: components/style/values/specified/mod.rs
Comment (pyfisch) [inline components/style/values/specified/mod.rs]: Yes this looks better.

**Comment by pyfisch** [comment]
@emilio fixed the other nits.

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: Probably not, I'm not sure that's defined right now, and if it is, we'd have to fix a bunch more, so it's fine. I think the old code clamped it there because we effectively didn't modify it afterwards, but clamping only on radians() seems reasonable.

File: components/style/values/specified/mod.rs
Comment (emilio) [inline components/style/values/specified/mod.rs]: Per the other conversation, this isn't really needed I guess, but I don't think it hurts, feel free to fix it up if you have the time before bors and treeherder report results.

Comment (pyfisch) [inline components/style/values/specified/mod.rs]: Done.

File: components/style/properties/longhand/box.mako.rs
Comment (canaltinova) [inline components/style/properties/longhand/box.mako.rs]: nit: use doc comments here as well?

**Review by canova — COMMENTED** [review]
It looks good to me too. Great work! I'll wait this PR to land for the -moz-transform then.

File: components/style/properties/longhand/box.mako.rs
Comment (pyfisch) [inline components/style/properties/longhand/box.mako.rs]: Yes. Done.

**Comment by emilio** [comment]
This PR exposes a few pre-existing bugs (like negative lengths being accepted for `perspective()`, but I don't see a reason to hold this back from landing, we can fix those in a follow-up.

Thanks a lot again for doing this work @pyfisch! :)

@bors-servo r+

**Comment by jdm** [comment]
```
  ▶ FAIL [expected PASS] /_mozilla/css/transform_skew_a.html
  └   → /_mozilla/css/transform_skew_a.html efb47d7f444001f095d2e342bc5a779238a22e0d
/_mozilla/css/transform_skew_ref.html e62b322bceb5c10cf1d0e0af232adcc29b52d2d7
Testing efb47d7f444001f095d2e342bc5a779238a22e0d == e62b322bceb5c10cf1d0e0af232adcc29b52d2d7
```
Reference:
screen shot 2017-04-03 at 3 04 45 pm

Non-reference:
screen shot 2017-04-03 at 3 04 07 pm

**Comment by GloomIndustries** [comment]
On Apr 3, 2017 3:05 PM, Josh Matthews <notifications@github.com> wrote:

  ▶ FAIL [expected PASS] /_mozilla/css/transform_skew_a.html
  └   → /_mozilla/css/transform_skew_a.html efb47d7f444001f095d2e342bc5a779238a22e0d
/_mozilla/css/transform_skew_ref.html e62b322bceb5c10cf1d0e0af232adcc29b52d2d7
Testing efb47d7f444001f095d2e342bc5a779238a22e0d == e62b322bceb5c10cf1d0e0af232adcc29b52d2d7


Reference:
[screen shot 2017-04-03 at 3 04 45 pm]<https://cloud.githubusercontent.com/assets/27658/24626319/e32e1a5e-187e-11e7-9b3e-c69a826ba22c.png>

Non-reference:
[screen shot 2017-04-03 at 3 04 07 pm]<https://cloud.githubusercontent.com/assets/27658/24626303/d41e9b9c-187e-11e7-8165-862ec762ef64.png>

**Comment by pyfisch** [comment]
@jdm I do not understand the test. Reference is what is expected and Non-reference is the rendering from my patch? I opened the file in Firefox and Chrome and both times I got the Non-reference rendering. [See for yourself](https://htmlpreview.github.io/?https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/css/transform_skew_a.html) Also servo renders it the same way (w/ and w/o patch). Am I using a wrong file or is there something else?

**Comment by jdm** [comment]
The point of reference tests is that there are two HTML files containing different content that are expected to produce identical rendered output. On Servo master, this test passes (ie. Servo renders both tests identically). With this PR applied, the output of the two files is different. "Reference" refers to `transform_skew_ref.html`, while "non-reference" refers to `transform_skew_a.html`. It would be worth checking the output of both files in other browsers to verify that Servo's behaviour on master is correct.

**Comment by jdm** [comment]
It appears that the changes in this PR affected the rendering of `transform_skew_ref.html` somehow.

File: components/style/properties/longhand/box.mako.rs
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: the bug is here: `comp.m12` should be `b`, not `c`. Similarly, `comp.m21` should be `c`, not `b`.

**Comment by pyfisch** [comment]
Fix pushed. Thanks to @emilio  for finding the bug.

**Comment by emilio** [comment]
@bors-servo r+

Thanks once again! :)

File: tests/wpt/metadata-css/css-transforms-1_dev/html/css-transform-style-evaluation-validation.htm.ini
Comment (emilio) [inline tests/wpt/metadata-css/css-transforms-1_dev/html/css-transform-style-evaluation-validation.htm.ini]: Instead of this, we can remove the ini file, since tests are pass-by-default.

Comment (pyfisch) [inline tests/wpt/metadata-css/css-transforms-1_dev/html/css-transform-style-evaluation-validation.htm.ini]: Ah ok, I will remove the file.

**Comment by pyfisch** [comment]
My conversion from matrix to computed matrix was still wrong. Hopefully the tests pass now.

**Comment by pyfisch** [comment]
@emilio can you please approve?

**Comment by pyfisch** [comment]
Travis failed because Ubuntu archives are currently unavailable.

**Comment by emilio** [comment]
Thanks for the awesome work here @pyfisch!

**Comment by pyfisch** [comment]
:smile:

> This PR exposes a few pre-existing bugs (like negative lengths being accepted for perspective(), but I don't see a reason to hold this back from landing, we can fix those in a follow-up.

What other bugs are exposed? Maybe I can write a follow-up.

**Comment by emilio** [comment]
See some of the failing tests at https://treeherder.mozilla.org/#/jobs?repo=try&revision=ecb364953429bc19871e35871dd2a7d9ae2d084f&selectedJob=88354219.

In particular, stuff like:

```
invalid value 'perspective(-10px)' not accepted for 'transform' property - got "perspective(-10px)", expected ""
```

Before this patch, we didn't serialize `perspective()` so this test was (incorrectly) passing. Now we serialize it, but accept negative values.

A test like:

```rust
<!doctype html>
<div style="transform: perspective(-10px)"></div>
<script>
alert(document.querySelector('div').style.transform)
</script>
```

Alerts the empty string on other browsers, it alerts `perspective(-10px)` with your patch. We should reject negative lengths when parsing the `perspective` function. Are you interested in working on that?

**Pull request merged**
